homeassistant/components/particle/__init__.py

```diff
@@ -0,0 +1,123 @@
+"""Integration with the Particle Cloud."""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_ACCESS_TOKEN, STATE_ON, STATE_UNAVAILABLE
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import make_entity_service_schema
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+
+DOMAIN = "particle"
+
+_LOGGER = logging.getLogger(DOMAIN)
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema({vol.Required(CONF_ACCESS_TOKEN): cv.string})},
+    extra=vol.ALLOW_EXTRA,
+)
+
+ATTR_ARGS = "args"
+ATTR_FUNCTION = "function"
+
+FUNCTION_CALL_SCHEMA = make_entity_service_schema(
+    {
+        vol.Required(ATTR_FUNCTION): cv.string,
+        vol.Optional(ATTR_ARGS): vol.All(
+            cv.ensure_list, vol.Length(min=1), [cv.string]
+        ),
+    }
+)
+
+
+async def async_setup(hass, config):
+    """Initialize our component while Home Assistant loads."""
+    if DOMAIN not in config:
+        return True
+
+    from pyparticleio.ParticleCloud import ParticleCloud
+
+    access_token = config[DOMAIN][CONF_ACCESS_TOKEN]
+
+    try:
+        particle = hass.data[DOMAIN] = ParticleCloud(access_token)
+    except Exception as ex:  # pylint: disable=broad-except
+        _LOGGER.error("Unable to connect to Particle Cloud: %s", str(ex))
+        return False
+
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
```

Comment:
Let's don't do this and instead create separate sensor entities in a sensor platform as needed for each attribute. We can load other platforms via discovery, discovery.async_load_platform. See other integrations for examples.

We can register a service here in the base integration module that calls the cloud function and takes the name of the device as service data.

Reply:
How's the documentation for doing this refactor? This has been working for quite some time, and it seems like there's been some internal, architectural changes that are affecting the way integrations are meant to be written.

It took me a long time to get this version working and approved, as:

- Updates take a long time to test, particularly for a DIY device like Particles.
- The desired output is a moving target.
- Documentation disagreed on what should be written where, and what documentation was considered "blessed" disagreed with the code already running on the contemporary versions of Home Assistant.

Your question earlier today was whether I intended to work through the review comments. I really do, but it's a lot to work through for a small, free contribution I wrote 6 months ago.

What I want has remained the same: I want some idea that when I'm done working through the current comments, I'll be done with this PR and it can be merged.

This started out simple enough: I had a custom component for a popular DIY platform, and I thought I'd contribute it to other Home Assistant users. If you have a moment to consider what this feels like for a newcomer, consider that every message I've received has included desired changes that conflict with what official documentation I could find, and some of which conflict with other reviews I've received on the same code.

So, once more from the top: where can I go to figure out what I _actually_ need to do here?

Reply:
Read our developer documentation and look at existing examples that are recently added. Older integrations are often not up to standard.

I can't give any guarantees on how much feedback there will be. We have standards that we want to maintain and we won't merge a PR until we think that it's ok according to those standards.

From yesterday we require all new integrations for devices and services to use a config flow and config entries. For the PRs where review has started we are making exceptions on this rule. That's why I haven't said anything about that until now.

I'd recommend reading the whole section about creating integrations that start with the file structure:
https://developers.home-assistant.io/docs/creating_integration_file_structure

Here's the docs on how to load a platform via discovery:
https://developers.home-assistant.io/docs/creating_component_generic_discovery#loading-platforms-when-configured-via-configurationyaml